File: 2024/Day_17/17.c
#include <aoc_tools.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

char *run(long long registers[3], int *program, int program_len) {
  int i = 0;
  char combos[] = {'0', '1', '2', '3', 'A', 'B', 'C'};
  char *out = malloc(40);
  out[0] = '\0';
  while (i < program_len) {
    int opcode = program[i], operand = program[i + 1];
    char char_combo_operand = combos[operand];
    long long combo_operand;
    if (isdigit(char_combo_operand)) {
      combo_operand = char_combo_operand - '0';
    } else {
      combo_operand = registers[char_combo_operand - 'A'];
    }

    switch (opcode) {
    case 0:
      registers[0] /= pow(2, combo_operand);
      break;
    case 1:
      registers[1] ^= operand;
      break;
    case 2:
      registers[1] = combo_operand % 8;
      break;
    case 3:
      if (registers[0] != 0) {
          i = operand;
          continue;
      }
      break;
    case 4:
      registers[1] ^= registers[2];
      break;
    case 5:
      char buffer[4];
      snprintf(buffer, sizeof(buffer), "%d,", combo_operand % 8);
      strcat(out, buffer);
      break;
    case 6:
      registers[1] = registers[0] / pow(2, combo_operand);
      break;
    case 7:
      registers[2] = registers[0] / pow(2, combo_operand);
      break;
    }

    i += 2;
  }
  return out;
}

int main(int argc, char *argv[]) {
  char *buffer = read_file(argc, argv);
  if (buffer == NULL) {
    return 1;
  }

  long long registers[3] = { 0, 0, 0 };
  char *line = strtok(buffer, "\n");
  sscanf(line, "Register A: %lld", &registers[0]);
  line = strtok(NULL, "\n");
  sscanf(line, "Register B: %lld", &registers[1]);
  line = strtok(NULL, "\n");
  sscanf(line, "Register C: %lld", &registers[2]);
  line = strtok(NULL, "\n");
  int line_len = strlen(line);
  int program_len = 0;
  int program_cap = 20;
  int *program = malloc(program_cap * sizeof(int));
  for (int c = 0; c < line_len; c++) {
    if (isdigit(line[c])) {
      if (program_len == program_cap) {
        program_cap *= 2;
        program = realloc(program, program_cap * sizeof(int));
      }
      program[program_len++] = line[c] - '0';
    }
  }

  char *p1 = run(registers, program, program_len);
  p1[strlen(p1) - 1] = '\0';

  long long p2 = 0;
  for (int i = program_len - 1; i > -1; i--) {
    char *p = malloc(40);
    p[0] = '\0';
    for (int j = i; j < program_len; j++) {
      char buffer[4];
      snprintf(buffer, sizeof(buffer), "%d,", program[j]);
      strcat(p, buffer);
    }
    p[strlen(p) - 1] = '\0';
    while (1) {
      //printf("%lld\n", p2);
      registers[0] = p2;
      char *out = run(registers, program, program_len);
      out[strlen(out) - 1] = '\0';
      //printf("%s\n", out);
      if (strcmp(out, p) == 0) {
        if (i != 0) {
          p2 *= 8;
        }
        break;
      }
      p2++;
      free(out);
    }
  }

  printf("%s\n", p1);
  printf("%lld\n", p2);

  free(buffer);
  free(program);
  free(p1);

  return 0;
}

File: 2024/Day_17/17.py
from aoc_tools import *

file = read_file()

registers, programs = file.split("\n\n")
registers = nums(registers)
programs = nums(programs)

combos = [0, 1, 2, 3, "A", "B", "C"]
register_loc = {"A": 0, "B": 1, "C": 2}

p1, p2 = "", 0


def run(registers, programs):
    combos = [0, 1, 2, 3, "A", "B", "C"]
    register_loc = {"A": 0, "B": 1, "C": 2}
    i = 0
    out = ""
    while i < len(programs):
        opcode, operand = programs[i], programs[i + 1]
        combo_operand = combos[operand]
        if not isinstance(combo_operand, int):
            combo_operand = registers[register_loc[combo_operand]]
        # print(registers, programs)
        # print(opcode, operand)
        # print(combo_operand)
        match opcode:
            case 0:
                registers[0] //= pow(2, combo_operand)
            case 1:
                registers[1] ^= operand
            case 2:
                registers[1] = combo_operand % 8
            case 3:
                if registers[0] != 0:
                    i = operand
                    continue
            case 4:
                registers[1] ^= registers[2]
            case 5:
                out += str(combo_operand % 8) + ","
            case 6:
                registers[1] = registers[0] // pow(2, combo_operand)
            case 7:
                registers[2] = registers[0] // pow(2, combo_operand)

        i += 2
    return out


registers_copy = registers.copy()
p1 = run(registers_copy, programs)

# 2 4 B = A % 8 (last 3 bits)
# 1 6 B = B ^ 110
# 7 5 C = A // 2**B
# 4 6 B = B ^ C
# 1 4 B = B ^ 100
# 5 5 out(B)
# 0 3 A = A // 8 (3 bits moved over)
# 3 0 jnz A

# 2,4,1,6,7,5,4,6,1,4,5,5,0,3,3,0
# 245 => 3, 3, 0
# 245322 => 550330
# 245325311 => 614550330
# 245325343540 => 754614550330
# 245325343540035 => 416754614550330
# 2453253435400351 => 2416754614550330

p2 = 0o2453253435400350
while True:
    registers_copy = registers.copy()
    registers_copy[0] = p2
    output = run(registers_copy, programs)
    if output.split(",")[:-1] == programs:
        break
    if output.startswith(
        ",".join(
            [
                "2",
                "4",
                "1",
                "6",
                "7",
                "5",
                "4",
                "6",
                "1",
                "4",
                "5",
                "5",
                "0",
                "3",
                "3",
                "0",
            ]
        )
    ):
        break
    p2 += 1
print(p1[:-1])
print(p2)
